POSIXFAIL_CLOCK_GETTIME second try

gperciva opened the pull request.

gperciva commented:
I'm not 100% confident about this yet, but I wanted to get a fresh PR up there after my git force-push mess.  (unfortunately github doesn't allow us to re-open a PR that was merged, even if the actual commits were removed)

You likely won't like the
```
#if THINGY
#      define USE_THINGY
#endif
```
portion.  I find it more readable, albeit a bit ugly.  Apparently really old compilers had to have the `#` in the left-most column but that's no longer required, but I figured I might as well keep them there for now.  I'm quite content to remove the spaces.

cperciva commented:
Yeah, not a fan of the indented preprocessor directives.

cperciva commented on POSIX/posix-cflags.sh:
```diff
@@ -21,6 +21,24 @@ if ! ${CC} -D_POSIX_C_SOURCE=200809L posix-clock_realtime.c 2>/dev/null; then
 	printf %s "-DPOSIXFAIL_CLOCK_REALTIME"
 	echo "WARNING: POSIX violation: <time.h> not defining CLOCK_REALTIME" 1>&2
 fi
+rm -f posix-clock_gettime
```
Why deleting the file here when we're about to write over it anyway?

gperciva replied:
It seemed more reliable that way.  What guarantees do we have about `c99` compilers faced with an error?

- if `gcc` (or `clang`) attempts to compile `/dev/null`, they delete `a.out`.
- if they fail because of a wrong header (or some random junk in the `.c` file), they don't touch `a.out`.

Granted, I suppose that we only run the binary if the compilation succeeds, and in that case, we're guaranteed that the output is fresh.  Hmm, ok.

cperciva commented on POSIX/posix-cflags.sh:
```diff
@@ -21,6 +21,24 @@ if ! ${CC} -D_POSIX_C_SOURCE=200809L posix-clock_realtime.c 2>/dev/null; then
 	echo "WARNING: POSIX violation: <time.h> not defining CLOCK_REALTIME" 1>&2
 fi
+rm -f posix-clock_gettime
+if ! ${CC} -D_POSIX_C_SOURCE=200809L -o posix-clock_gettime posix-clock_gettime.c 2>/dev/null; then
```
... and for that matter, why not use let it have the default `a.out` name?

gperciva commented:
Updated PR to keep the default `a.out` binary name.  Tested on OSX 10.11.

Since you're distracted by a conference and stuff, I went ahead and rebased the history.

cperciva commented:
Hmm, thought which just occurred to me... is it possible that the OS X issue is caused by `clock_gettime` being in `librt`?

gperciva commented:
The original problem reporter wrote:
> clock_gettime(3) is found in libc (/usr/lib/system/libsystem_c.dylib) on macOS 10.12+. There's no seperate librt
https://github.com/Tarsnap/tarsnap/issues/286#issuecomment-323894105

I could dig around for more references for this... (grr, I wish that we could virtualize OSX!  I only have one osx machine, and I don't want to upgrade it to 10.12 while we still need to test 10.11 stuff.)

cperciva commented:
Thanks, I remember the history now.

cperciva commented:
Changes look good, but

```
OSX 10.11 and below did not include clock_gettime(), despite it being a
mandatory part of POSIX.
```

To avoid future confusion, can you add "To make things worse, clock_gettime and CLOCK_REALTIME are defined in <time.h> despite the symbol not being exported by the system libraries; so we have to detect this case by trying to run code which uses clock_gettime rather than relying on whether preprocessor macros are defined."

gperciva commented:
Rebased on master, changing both the comment in `posix-cflags.sh` and the commit message.

gperciva commented:
FWIW, here's the results with the basic 10.11 system.  Note the different POSIX violation messages.
```
Grahams-Mac-mini:monoclock gperciva$ make
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f Makefile.inc1 CFLAGS_POSIX="`export CC="cc"; cd ../../POSIX && command -p sh posix-cflags.sh "$PATH"`" LDADD_POSIX="`export CC="cc"; cd ../../POSIX && command -p sh posix-l.sh "$PATH"`" CFLAGS="-I/usr/local/opt/openssl/include " test_monoclock
WARNING: POSIX violation: /bin/bash's command -p resets $PATH
WARNING: POSIX violation: <sys/socket.h> not defining MSG_NOSIGNAL
WARNING: POSIX violation: <time.h> not defining CLOCK_REALTIME
WARNING: POSIX violation: <time.h> not declaring clock_gettime()
WARNING: POSIX violation: /bin/bash's command -p resets $PATH
WARNING: POSIX violation: make's CC doesn't understand -lrt
WARNING: POSIX violation: make's CC doesn't understand -lxnet
cc -I/usr/local/opt/openssl/include  -D_POSIX_C_SOURCE=200809L -I ../../util -DPOSIXFAIL_MSG_NOSIGNAL -DPOSIXFAIL_CLOCK_REALTIME -DPOSIXFAIL_CLOCK_GETTIME -D__BSD_VISIBLE -c main.c -o main.o
cc -I/usr/local/opt/openssl/include  -D_POSIX_C_SOURCE=200809L -I ../../util -DPOSIXFAIL_MSG_NOSIGNAL -DPOSIXFAIL_CLOCK_REALTIME -DPOSIXFAIL_CLOCK_GETTIME -c ../../util/monoclock.c -o monoclock.o
cc -I/usr/local/opt/openssl/include  -D_POSIX_C_SOURCE=200809L -I ../../util -DPOSIXFAIL_MSG_NOSIGNAL -DPOSIXFAIL_CLOCK_REALTIME -DPOSIXFAIL_CLOCK_GETTIME -c ../../util/warnp.c -o warnp.o
cc -I/usr/local/opt/openssl/include  -o test_monoclock main.o monoclock.o warnp.o 
Grahams-Mac-mini:monoclock gperciva$ ./test_monoclock && echo $?
0
Grahams-Mac-mini:monoclock gperciva$ 
```

and here's the result with 10.11 + Xcode 8 (so 10.12 SDK):

```
Grahams-Mac-mini:monoclock gperciva$ more ~/break-clock-gettime.sh 
export SDKROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk
export MACOSX_DEPLOYMENT_TARGET=10.11
Grahams-Mac-mini:monoclock gperciva$ source ~/break-clock-gettime.sh 
Grahams-Mac-mini:monoclock gperciva$ make
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f Makefile.inc1 CFLAGS_POSIX="`export CC="cc"; cd ../../POSIX && command -p sh posix-cflags.sh "$PATH"`" LDADD_POSIX="`export CC="cc"; cd ../../POSIX && command -p sh posix-l.sh "$PATH"`" CFLAGS="-I/usr/local/opt/openssl/include " test_monoclock
WARNING: POSIX violation: /bin/bash's command -p resets $PATH
WARNING: POSIX violation: <sys/socket.h> not defining MSG_NOSIGNAL
WARNING: POSIX violation: clock_gettime() is not linkable
WARNING: POSIX violation: /bin/bash's command -p resets $PATH
WARNING: POSIX violation: make's CC doesn't understand -lrt
WARNING: POSIX violation: make's CC doesn't understand -lxnet
cc -I/usr/local/opt/openssl/include  -D_POSIX_C_SOURCE=200809L -I ../../util -DPOSIXFAIL_MSG_NOSIGNAL -DPOSIXFAIL_CLOCK_GETTIME -D__BSD_VISIBLE -c main.c -o main.o
cc -I/usr/local/opt/openssl/include  -D_POSIX_C_SOURCE=200809L -I ../../util -DPOSIXFAIL_MSG_NOSIGNAL -DPOSIXFAIL_CLOCK_GETTIME -c ../../util/monoclock.c -o monoclock.o
cc -I/usr/local/opt/openssl/include  -D_POSIX_C_SOURCE=200809L -I ../../util -DPOSIXFAIL_MSG_NOSIGNAL -DPOSIXFAIL_CLOCK_GETTIME -c ../../util/warnp.c -o warnp.o
cc -I/usr/local/opt/openssl/include  -o test_monoclock main.o monoclock.o warnp.o 
Grahams-Mac-mini:monoclock gperciva$ ./test_monoclock && echo $?
0
```
